iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs

```diff
@@ -258,88 +258,104 @@ private void DisposeDelivery(AmqpMessage amqpMessage, bool settled, Accepted acc
+                    }
+                    else
+                    {
+                        throw new IotHubException(error, false);
+                    }
+                }
+                else
                 {
-                    if (amqpMessage.BodyStream != null)
+                    Twin twin = null;
+                    TwinCollection twinProperties = null;
+
+                    if (correlationId == null)
                     {
-                        // This a result of a GET TWIN so return (set) the full twin
+                        // Here we are getting desired property update notifications and want to handle it first
+                        using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                        {
+                            string patch = reader.ReadToEnd();
+                            twinProperties = JsonConvert.DeserializeObject<TwinCollection>(patch);
+                        }
+                    }
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Get.ToString(), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // This a response of a GET TWIN so return (set) the full twin
                         using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
                         {
                             string body = reader.ReadToEnd();
                             var properties = JsonConvert.DeserializeObject<TwinProperties>(body);
                             twin = new Twin(properties);
                         }
                     }
-                    else
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Patch.ToString(), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // This can be used to coorelate success response with Updating reported properties
+                    }
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Put.ToString(), StringComparison.OrdinalIgnoreCase))
                     {
-                        // This is a desired property ack from the service
+                        // This is an acknowledgement received from service for subscribing to desired property updates
                         twin = new Twin();
                     }
-                }
-                else
-                {
-                    // No correlationId, this is a PATCH sent by the sevice so return (set) the TwinCollection
-
-                    using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                    else
                     {
-                        string patch = reader.ReadToEnd();
-                        twinProperties = JsonConvert.DeserializeObject<TwinCollection>(patch);
+                        // This shouldn't happen
```

Comment:
Maybe we need to log a warning here in case it ever happens.

Reply:
... or it should never ever happen? :)

Reply:
Ideally it shouldn't but you are right adding a log is a better idea anyways.

Reply:
+1 to adding log warning here, if relevant